scylla-cdc/src/stream_reader.rs

```diff
@@ -134,17 +139,44 @@ impl StreamReader {
             values[begin_index] = CqlValue::Timestamp(window_begin);
             values[end_index] = CqlValue::Timestamp(window_end);
 
-            let mut rows_stream = self
-                .session
-                .execute_iter(query_base.clone(), &values)
-                .await?;
-
-            let schema = CDCRowSchema::new(rows_stream.get_column_specs());
-
-            while let Some(row) = rows_stream.next().await {
-                consumer
-                    .consume_cdc(CDCRow::from_row(row?, &schema))
-                    .await?;
+            let mut sleep_after_timeout = BASIC_TIMEOUT_SLEEP;
+
+            let mut next_state = None;
+            'try_query: loop {
+                let state_clone = next_state.clone();
+                let query_res = self
+                    .session
+                    .execute_paged(&query_base, &values, next_state)
+                    .await;
+                match query_res {
+                    Ok(x) => {
+                        sleep_after_timeout = BASIC_TIMEOUT_SLEEP;
+                        if let Some(rows) = x.rows {
+                            let schema = CDCRowSchema::new(&x.col_specs);
+
+                            for row in rows {
+                                consumer
+                                    .consume_cdc(CDCRow::from_row(row, &schema))
+                                    .await?;
+                            }
+                        }
+                        match x.paging_state {
+                            Some(state) => next_state = Some(state),
+                            None => break 'try_query,
+                        }
+                    },
+                    Err(QueryError::TimeoutError) => {
```

Comment:
This error is for client side timeouts (AFAIK they are not implemented for regular queries yet, but it's OK if you handle this case explicitly now). You also need to handle server-side timeouts: https://github.com/scylladb/scylla-rust-driver/blob/9d402a92558485ee8343386c4204d9c1d6ca19c0/scylla/src/transport/errors.rs#L124